crates/bevy_ecs/src/world/mod.rs

```diff
@@ -0,0 +1,819 @@
+    /// position.x = 1.0;
+    /// ```
+    #[inline]
+    pub fn entity_mut(&mut self, entity: Entity) -> EntityMut {
+        self.get_entity_mut(entity).expect("Entity does not exist")
+    }
+
+    /// Retrieves an [EntityRef] that exposes read-only operations for the given `entity`.
+    /// Returns [None] if the `entity` does not exist. Use [World::entity] if you don't want
+    /// to unwrap the [EntityRef] yourself.
+    ///
+    /// ```
+    /// use bevy_ecs::world::World;
+    ///
+    /// struct Position {
+    ///   x: f32,
+    ///   y: f32,
+    /// }
+    ///
+    /// let mut world = World::new();
+    /// let entity = world.spawn()
+    ///     .insert(Position { x: 0.0, y: 0.0 })
+    ///     .id();
+    ///
+    /// let entity_ref = world.get_entity(entity).unwrap();
+    /// let position = entity_ref.get::<Position>().unwrap();
+    /// assert_eq!(position.x, 0.0);
+    /// ```
+    #[inline]
+    pub fn get_entity(&self, entity: Entity) -> Option<EntityRef> {
+        let location = self.entities.get(entity)?;
+        Some(EntityRef::new(self, entity, location))
+    }
+
+    /// Retrieves an [EntityMut] that exposes read and write operations for the given `entity`.
+    /// Returns [None] if the `entity` does not exist. Use [World::entity_mut] if you don't want
+    /// to unwrap the [EntityMut] yourself.
+    ///
+    /// ```
+    /// use bevy_ecs::world::World;
+    ///
+    /// struct Position {
+    ///   x: f32,
+    ///   y: f32,
+    /// }
+    ///
+    /// let mut world = World::new();
+    /// let entity = world.spawn()
+    ///     .insert(Position { x: 0.0, y: 0.0 })
+    ///     .id();
+    ///
+    /// let mut entity_mut = world.get_entity_mut(entity).unwrap();
+    /// let mut position = entity_mut.get_mut::<Position>().unwrap();
+    /// position.x = 1.0;
+    /// ```
+    #[inline]
+    pub fn get_entity_mut(&mut self, entity: Entity) -> Option<EntityMut> {
+        let location = self.entities.get(entity)?;
+        // SAFE: `entity` exists and `location` is that entity's location
+        Some(unsafe { EntityMut::new(self, entity, location) })
+    }
+
+    /// Spawns a new [Entity] and returns a corresponding [EntityMut], which can be used
+    /// to add components to the entity or retrieve its id.
+    ///
+    /// ```
+    /// use bevy_ecs::world::World;
+    ///
+    /// struct Position {
+    ///   x: f32,
+    ///   y: f32,
+    /// }
+    ///
+    /// let mut world = World::new();
+    /// let entity = world.spawn()
+    ///     .insert(Position { x: 0.0, y: 0.0 }) // add a single component
+    ///     .insert_bundle((1, 2.0, "hello")) // add a bundle of components
+    ///     .id();
+    ///
+    /// let position = world.entity(entity).get::<Position>().unwrap();
+    /// assert_eq!(position.x, 0.0);
+    /// ```
+    pub fn spawn(&mut self) -> EntityMut {
+        self.flush();
+        let entity = self.entities.alloc();
+        let archetype = self.archetypes.empty_mut();
```

Comment:
This can use `.empty()`, right?

Reply:
Nope! `archetype.allocate()` requires a mutable reference.